File: VB/App_Code/BasePage.vb
Option Infer On

Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Web
Imports System.Web.Services

Public Class BasePage
    Inherits System.Web.UI.Page

    Private Shared ReadOnly Property Operation() As Operation
        Get
            If HttpContext.Current.Session("_Operation") Is Nothing Then
                HttpContext.Current.Session("_Operation") = New Operation()
            End If
            Return TryCast(HttpContext.Current.Session("_Operation"), Operation)
        End Get
    End Property

    <WebMethod(EnableSession := True)> _
    Public Shared Function GetProgress() As String
        If HttpContext.Current.Session Is Nothing OrElse Not (TypeOf HttpContext.Current.Session("_Operation") Is Operation) Then
            Return "NaN"
        End If
        Dim progress = (TryCast(HttpContext.Current.Session("_Operation"), Operation)).Progress
        Return progress.ToString()
    End Function

    <WebMethod(EnableSession := True)> _
    Public Shared Sub StartOperation()
        Operation.AsyncStart(5000)
    End Sub
End Class

File: CS/App_Code/BasePage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web;
using System.Web.Services;

public class BasePage : System.Web.UI.Page {
    static Operation Operation {
        get {
            if (HttpContext.Current.Session["_Operation"] == null)
                HttpContext.Current.Session["_Operation"] = new Operation();
            return HttpContext.Current.Session["_Operation"] as Operation;
        }
    }

    [WebMethod(EnableSession = true)]
    public static string GetProgress() {
        if (HttpContext.Current.Session == null || !(HttpContext.Current.Session["_Operation"] is Operation))
            return "NaN";
        var progress = (HttpContext.Current.Session["_Operation"] as Operation).Progress;
        return progress.ToString();
    }

    [WebMethod(EnableSession = true)]
    public static void StartOperation() {
        Operation.AsyncStart(5000);
    }
}
